### PR TITLE
Support both file_secret and secret_id based paths

### PR DESCRIPTION
**Pull request by mephenor**
Old code paths and related classes dealing with output_metadata including the plain text file secret are now prefixed with legacy.
Refactored s3_upload into a subpackage to keep some semblance of sanity.
Added missing object storage cleanup when checksum validation fails after re-downloading file for validation. 
Added secret_ingest URL and public key options to support file secret exchange directly after upload.
Token for communication on that path is read from the same location as for the file ingest functionality before, i.e. the same token is used for both new FIS endpoints.

Please tell me if there are some tests missing that should be included, lost quite a bit of time here and probably missed one or two smaller issues.

As this repo does not track the microservice template, we probably should also have a refactoring backlog entry to move this to ruff, pydantic v2 and our pip lock setup.

### CODE REVIEW

File: ghga_datasteward_kit/s3_upload/config.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import subprocess  # nosec
+from pathlib import Path
+
+from pydantic import BaseSettings, Field, SecretStr, validator
+
+
+def expand_env_vars_in_path(path: Path) -> Path:
+    """Expand environment variables in a Path."""
+
+    with subprocess.Popen(  # nosec
```
Comment (Cito) [inline ghga_datasteward_kit/s3_upload/config.py:27]: Maybe easier to use subprocess.run here.

File: ghga_datasteward_kit/s3_upload/downloader.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    async def validate_checksums(self, checkums: models.Checksums):
+        """Confirm checksums for upload and download match"""
+        if not self.target_checksums.get() == checkums.get():
+            object_storage = objectstorage(config=self.config)
+            await object_storage.delete_object(
+                bucket_id=self.config.bucket_id, object_id=self.file_id
+            )
+            raise ValueError(
+                f"Checksum mismatch:\nUpload:\n{checkums}\nDownload:\n{self.target_checksums}"
+                + "Uploaded file was deleted due to validation failure."
```
Comment (Cito) [inline ghga_datasteward_kit/s3_upload/downloader.py:88]:
```suggestion
                "Checksum mismatch:\n"
                + f"Upload:\n{checkums}\nDownload:\n{self.target_checksums}\n"
                + "Uploaded file was deleted due to validation failure."
```

One newline was missing.

File: ghga_datasteward_kit/s3_upload/entrypoint.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        )
+
+        if response.status_code != 200:
+            object_storage = objectstorage(config=config)
+            await object_storage.delete_object(
+                bucket_id=config.bucket_id, object_id=file_id
+            )
+            raise ValueError(
+                f"Failed to deposit secret for {alias} with response code"
+                + f"{response.status_code}. Removed uploaded file."
```
Comment (Cito) [inline ghga_datasteward_kit/s3_upload/entrypoint.py:117]:
```suggestion
                + f" {response.status_code}. Removed uploaded file."
```

**Review by Cito — APPROVED** [review]
Could not really check in depth.

**Review by Cito — DISMISSED** [review]

File: ghga_datasteward_kit/s3_upload/entrypoint.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    check_adjust_part_size,
+    check_output_path,
+    handle_superficial_error,
+    httpx_client,
+    objectstorage,
+)
+from ghga_datasteward_kit.utils import load_config_yaml, read_token
+
+
+async def shared_path(input_path: Path, alias: str, config: LegacyConfig):
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/entrypoint.py:45]: Return type not specified. Maybe rename and make docstring more specific.

File: ghga_datasteward_kit/s3_upload/entrypoint.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    with httpx_client() as client:
+        headers = {"Authorization": f"Bearer {token}"}
+        response = client.post(
+            url=config.secret_ingest_url,
+            json=encrypted_secret.dict(),
+            headers=headers,
+            timeout=60,
+        )
+
+        if response.status_code != 200:
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/entrypoint.py:110]: I wouldn't see the cleanup of the bucket as the responsibility of this function. However, if you cannot find an easy alternative, this is also fine.

File: ghga_datasteward_kit/s3_upload/entrypoint.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        return response.json()["secret_id"]
+
+
+async def async_main(input_path: Path, alias: str, config: Config, token: str):
+    """
+    Run encryption, upload and validation.
+    Prints metadata to <alias>.json in the specified output directory
+    """
+
+    uploader, file_size = await shared_path(
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/entrypoint.py:128]: Maybe in general, a context manager at this level that just cleans up when something goes wrong would be an option. What do you think?

File: ghga_datasteward_kit/s3_upload/file_decryption.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/file_decryption.py:1]: Were the file decrypter, encrypte, and uploader rewritten? Or is GitHub just not able to detect that they where moved into different files? I.e. do I have to do a thorough review of them?

File: ghga_datasteward_kit/s3_upload/utils.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    while True:
+        file_part = file.read(part_size)
+
+        if len(file_part) == 0:
+            return
+
+        yield file_part
+
+
+def objectstorage(config: LegacyConfig):
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/utils.py:82]:
```suggestion
def get_objectstorage(config: LegacyConfig):
```

File: ghga_datasteward_kit/s3_upload/utils.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    return byte_ranges
+
+
+def handle_superficial_error(msg: str):
+    """Don't want user dealing with stacktrace on simple input/output issues, log instead"""
+    LOGGER.critical(msg)
+    sys.exit(-1)
+
+
+def check_adjust_part_size(config: LegacyConfig, file_size: int):
```
Comment (KerstenBreuer) [inline ghga_datasteward_kit/s3_upload/utils.py:122]: Is this new or just moved here?

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Wow, this was a massive rewrite. However, I could just do a shallow review but did not found major issues. Would need much more time to go into more depth. Could you maybe help by highlighting which parts were rewritten and which parts mostly moved into different files?

Regarding tests, since this is becoming a quite massive codebase that is difficult to keep an overview (and to help answer your question about whether everything was tested correctly), maybe it would be good to try to approach 100% coverage. Some parts really do not need coverage, you can ignore them by configuring the coverage tools.

What do you think, would 100% coverage give you a better feel?

File: ghga_datasteward_kit/file_ingest.py [hunk 10]
```diff
@@ -120,13 +124,20 @@ def file_ingest(
 
     with httpx.Client() as client:
         response = client.post(
-            f"{config.file_ingest_url}", json=encrypted.dict(), headers=headers
+            f"{config.file_ingest_url}",
```
[the remark this answers is not in the document]
Comment (mephenor) [reply ghga_datasteward_kit/file_ingest.py:127]: Probably a good Idea, would go well with removing the legacy_mode flag

File: ghga_datasteward_kit/s3_upload/entrypoint.py (hunk 4 above)
Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:45]: What would you rename this to?

File: ghga_datasteward_kit/s3_upload/entrypoint.py (hunk 5 above)
Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:110]: See next comment

File: ghga_datasteward_kit/s3_upload/entrypoint.py (hunk 6 above)
Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:128]: Currently there's a context manager for the multipart upload that also deals with cleanup in that context.
Moving the cleanup functionality into a dedicated context manager on a higher level would make sense, but the problem here is getting the upload ID to the correct place, which is only generated on initializing the multipart upload.

Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:128]: Probably could do that using a an attribute on a custom exception, I'll see if I can come up with something simple

File: ghga_datasteward_kit/s3_upload/file_decryption.py (hunk 7 above)
Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/file_decryption.py:1]: It's a git issue. From everything that's moved into the s3_upload path, only config and entrypoint should have real changes.

File: ghga_datasteward_kit/s3_upload/utils.py (hunk 9 above)
Comment (mephenor) [reply ghga_datasteward_kit/s3_upload/utils.py:122]: Moved

File: ghga_datasteward_kit/s3_upload/entrypoint.py (hunk 6 above)
Comment (KerstenBreuer) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:128]: Sounds good. But no worries if it becomes to difficult.

File: ghga_datasteward_kit/s3_upload/file_decryption.py (hunk 7 above)
Comment (KerstenBreuer) [reply ghga_datasteward_kit/s3_upload/file_decryption.py:1]: OK, thanks for the clarification.

File: ghga_datasteward_kit/s3_upload/entrypoint.py (hunk 4 above)
Comment (KerstenBreuer) [reply ghga_datasteward_kit/s3_upload/entrypoint.py:45]: Maybe 
```suggestion
async def check_and_upload_content(input_path: Path, alias: str, config: LegacyConfig):
    """
    Check and upload encrypted file content. This also includes a verification of the upload by
    download the content again and perform a checksum validation.
    
    Returns:
        A tuple of the used uploader instance and the file size
    """
```

File: ghga_datasteward_kit/file_ingest.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mephenor) [reply ghga_datasteward_kit/file_ingest.py:127]: I also adjusted the secret_ingest_url to be the base URL and factored url composition out to the top of all affected functions.
The endpoints are hardcoded, which should suffice for the current use case.

**Comment by mephenor** [comment]
All requested changes have been addressed in the last three commits.
Instead of dealing directly with object storage cleanup a StorageCleaner context manager is passed through.
Two error case that previously did not lead to the deletion of an unusable uploaded file are now also covered: Failures during file part download  and writing the corresponding output file.

**Review by KerstenBreuer — APPROVED** [review]
Very nice thank you

**Pull request merged**